### ISSUE TITLE
resource-hwloc does not offer aggregate resource counts across ranks

### ISSUE THREAD

**Issue by trws** [flux-framework/flux-core#497]
Working on ATS, I find it would be useful to have a top-level aggregate for things like number of cores, PUs, etc. available to the scheduler as a whole.

@dongahn where did you end up on reduction mechanisms? That would seem like the right way to set this up.

**Comment by dongahn** [comment]
@trws: 

Aggregation seems like a great effort! 

Before I realized I wanted to spend more time on sched, I was looking at the `kvs_reduce` -- named reduced. I have a prototype atop of the older version of native reduce. The new interface is available at [here](https://github.com/flux-framework/flux-core/blob/bdde42af9d6e5d25287acccde9841b4a26f2f11d/src/common/libflux/reduce.h)

I don't think my prototype is tested and stable enough to be used by this effort. But at the same time I can see your case would be a perfect one that can use the named reduce because your case would use reduction as a collective (e.g., known participants) with either known or custom reduction operator.  

If this work can wait until early next year, it feels using your efforts to shape up the 'kvs_reduce' would be extremely helpful. If this can't wait, you should consider directly go to the native reduce.h support. IMO, the native interface is also pretty flexible to use.

**Comment by garlick** [comment]
See `flux_reduce_create(3)` though it may be a bit off target for this use.

If you're already pushing all the unreduced data to the KVS, is the added complexity worth it?  Just have the rank 0 resource-hwloc generate the reduced data once the fence on the unreduced data completes?

**Comment by trws** [comment]
That's probably easiest, and it's what my ATS python code does at the moment, but it's not going to scale.  You're probably right that it isn't worth it yet anyway though.

**Comment by dongahn** [comment]
FWIW, this is a general comment. My experiences with scalable tools tell me that we will probably benefit from reduction more at some point to reach the scale we need. What we learned was, at extreme scale, even aggregation that works as concatenation doesn't even cut it: we needed a good reduction that can actually reduce the size of the aggregated object.  

Some papers in case you are interested in the lessons: STAT merge time issue ([SC08-Lee-STAT.pdf](https://github.com/flux-framework/flux-core/files/64451/SC08-Lee-STAT.pdf)) and why we used bloom filters to avoid variable-length reduction ([FGFS-IPDPS13-4971a369-proceedings.pdf](https://github.com/flux-framework/flux-core/files/64455/FGFS-IPDPS13-4971a369-proceedings.pdf))

But I agree. We are probably not there yet.
